### ISSUE TITLE
Loosing token

### ISSUE THREAD

**Issue by barabba9174** [aholstenson/miio#206]
After the latest firmware update, I loose the token when I try to setup the wireless both from code and from app.

```
miio discover --sync

Device ID: 47119176
Model info: zhimi.airpurifier.m1
Address: 192.168.13.1
Token: 78235871355dc262387a9a6ecef3b687 via auto-token
Support: At least basic
```

And everything is fine for now... 
I try to change the wireless by device id:

```
miio configure 47119176 --ssid myssid --passwd mypassword
Attempting to configure 47119176
Could not find device
```
If I try` miio discover`  and the device is found correctly 

I try to change the wireless by ip:
```
miio configure 192.168.13.1 --ssid myssid --passwd mypassword
Attempting to configure 192.168.13.1
Encountered an error while updating wireless
Error was: Failed updating wireless
```

Trying again and the error changes...

```
miio configure 192.168.13.1 --ssid myssid --passwd mypassword
Attempting to configure 192.168.13.1
Error was: Encountered an error while updating wireless
Could not connect to device, token needs to be specified
```

and then 

```
miio discover

Device ID: 47119176
Model info: Unknown
Address: 192.168.13.1
Token: ???
Support: Unknown

```
Do you have any idea about the reason?

Also updating manually the token doesn't work

**Comment by silvetti** [comment]
I am having the same issue after latest firmware update :(

**Comment by aatishb** [comment]
I'm encountering the same issue. Also, when I try to inspect the device, I get the following error:
`ERROR  Could inspect device. Error was: Could not connect to device, token needs to be specified`

**Comment by findemor** [comment]
Same here,

I got the token using miio (with --sync param), 

```
Device ID: 131148119
Model info: yeelink.light.color2
Address: 192.168.13.1
Token: 9c2c7995a88583f7b8dd10a69e8f963f via auto-token
Support: At least generic
```

and when I connect the device to my Wifi Network using Mi HOme app, and when I try to discover devices the Token is null again and I can't connect to device.

When I try:

`miio.device({ address: device.address, token: "9c2c7995a88583f7b8dd10a69e8f963f" })`

I get:

```
code:"connection-failure"
device:null
message:"Could not connect to device, token might be wrong"
```

**Comment by maslowivan** [comment]
Check this issue, maybe it can help https://github.com/aholstenson/miio/issues/92

**Comment by acchavezaiot** [comment]
I have the same issue executing this command :  `miio protocol call 192.168.x.x get_prop '["temperature"]'` in Xiaomi MIJIA Air Quality Monitor Air Detector , i force to set the token in the file call.js  with this line `device.management.api.parent.packet._token = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";` but the command get the same issue `Could not connect to device, token needs to be specified`. someone has a solution or the miio package for new devices is deprecated? @aholstenson

**Comment by sachaw** [comment]
Very similar issue, token from discovery does not work after setup.

**Comment by phrogg** [comment]
Yeah same issue with the ceiling light 10 (Crystal Pendant Lamp). After changing the Network with this error message: "ERROR  Encountered an error while updating wireless", I can see the lamp connecting to my network correctly but it seems like it generated a new token because the old one is not accepted anymore...

**Comment by dazhbog** [comment]
I seem to experience the same issue. 

Connecting to the device's hotspot, the token is always available. Once the device changes WiFi the token seems to change (and prob uploaded to xiaomi through WiFi and downloaded from the app)

**Comment by stbenjam** [comment]
Same, miio works fine when connected to the wifi from the device, but after configuring it to talk to a network, the token no longer works. It seems to change.

**Comment by dazhbog** [comment]
@stbenjam Yes, I posted a solution here https://github.com/aholstenson/miio/issues/93#issuecomment-531206729

Not elegant but I was able to see that indeed the token is changing on network change.

**Comment by phrogg** [comment]
@dazhbog Yeah that's a nice workaround, however for me the whole point here is, not to use the app at all.

**Comment by stbenjam** [comment]
Yea it's not really a solution for me when it's an entirely different process.  I did manage to dig out an ancient android phone and recover the password from the database, but I'd rather not have to do that again if the device ever gets reset.

Hopefully someone figures out how to get the token with miio.

**Comment by dazhbog** [comment]
@phrogg That was why I also bought into this mi system too. I saw that miio "just works".

But its sad that Xiaomi is trying to stop people from accessing their devices this way. Btw I wrote a little program to try and reverse engineer the whole thing. Similar to miio, but i was able to play with the raw packets more. After playing with it for 2 days, it seems that the device resets its token if it can talk to xiaomi servers directly. (normally after changing wifi)

Using the app way to set up the device and then get the token u can have both the app and miio working together which is nice.

**Comment by phrogg** [comment]
@dazhbog Yeah.

When someone needs a convinient way to extract the tokens from the yeelight app https://github.com/phrogg/GetMiTokens is an App for Android to help with this specific task. Also available in F-Droid and PlayStore.

**Comment by mitchins** [comment]
Hi all, I replied in #93 [comment](https://github.com/aholstenson/miio/issues/93#issuecomment-568229827)
It may not be suitable for everyone, but if you either quarantine the mi device entirely from the internet, or even just blacklist the specific servers somehow (in theory, did not test) it won't change the token and API can be used after device joining a managed network (AP).

**Comment by phrogg** [comment]
@mitchins I will give that a try, when I get back home in the new year. Will reply here then.

**Comment by bdruth** [comment]
I can't get a plug to even use my wifi via `miio` - works fine from MiHome, but then with the aforementioned problems (I don't have a good way of blocking it from Internet access, unfortunately). Connecting to the plug's WiFi, `discover` works, but `configure` always gives an ~~error~~ warning that no devices were configured.

I wonder if the API to set the WiFi info has been changed as well for some devices?

**Comment by mitchins** [comment]
How about adding a $5 RTL8188EU to your pi or HASS host and using hostAPD?

**Comment by daweizhangau** [comment]
I have the same issue with my `chuangmi.plug.v3` Power Plug. It has the most recent firmware `2.0.4_0026`. The token changes after connecting it to WiFi. It changes even the device's blocked from accessing Internet.

I obtain the correct token by using the sqlite DB from Mi Home app for iOS ([see](https://python-miio.readthedocs.io/en/latest/discovery.html#tokens-from-backups)). Tried it twice, got the right token at the second trial due to some reason (after the app shows it's a NEW device).

**Comment by mitchins** [comment]
> I have the same issue with my `chuangmi.plug.v3` Power Plug. It has the most recent firmware `2.0.4_0026`. The token changes after connecting it to WiFi. It changes even the device's blocked from accessing Internet.
> 
> I obtain the correct token by using the sqlite DB from Mi Home app for iOS ([see](https://python-miio.readthedocs.io/en/latest/discovery.html#tokens-from-backups)). Tried it twice, got the right token at the second trial due to some reason (after the app shows it's a NEW device).

That seems strange and is definitely NOT what I encountered. Are you sure it did not have internet access somehow? I'd have to check the specific firmware version of my purifier but it DOES change when connected to the internet. How are you blocking internet access? I actually subnet it and don't give it a route out.
I'm pretty sure the change of token is actually an exchange of token with Xiaomi's servers because the most modern versions of the app no longer keep the token.
I also used command line miiocli to setup wifi, maybe the app does something else or maybe the older app with embedded SQLite store exchanges directly with the device? (as in perhaps the latest firmware mutates the token and sends it to cloud or legacy app as appropriate).

Sorry for the ramble, in short:
1. Try ensuring it definitely has no internet access (internet connection sharing and hotspots will by default, and 'blacklists' may always let one escape)
2. Use the command line tools from a computer to extract token and configure wifi.

**Comment by dzhang-quest** [comment]
@mitchins 
I tried the blacklist and I unplug the cable from the Router the device's connecting. So for sure, the later one will block Internet access. 

1. - Used command line to obtain the token `miplug discover`, 
2. - tested the token works, by check device status `miplug --ip 192.168.4.1 --token <your_token> -d status` (it's direct connection)
3. - Had the device connect to my router `miio configure 192.168.4.1 --ssid <ssid> --passwd <pwd>`
4. - Connected to the router from my workstation, and ping it pass. 
5. - `miplug --ip 192.168.10.149 --token <your_token> -d status` complained, No response.

Obtained the token from SQLite DB of the app, and running the command of Step 5 worked.

**Comment by tarapon** [comment]
I also can confirm that my Air humidifier also changes token after connect to wifi. Even if it’s blocked from internet. 

Actually when it’s blocked from the internet it can not be paired even with mi home app.

**Comment by mitchins** [comment]
Curious, I didn't use the app as mentioned I used the CLI config.

**Comment by poofyteddy** [comment]
I'm having the same issue with my new roborock S6.
I could get the token just fine from the AP created by the robot, but a soon as i setup a wifi, both with the roborock app and the cli, the token is nolonger valid.

I'm kinda face to a wall right now, i can't get the token that way, i don't have a rooted device, and i don't want to use the xiaomi app who ask for all privilege. I'm ok with the roborock app but it don't seam to have any flaw to get the token.

Is there any other way ? I still have 10D to ask for a refund, i think i'll need it.

**Comment by dzhang-quest** [comment]
@poofyteddy Have you tried to obtain the token from SQLite DB of the mobile app?

**Comment by poofyteddy** [comment]
@dzhang-quest sorry but, wich app are you talking about ?
I know the xiaomi mija app this, but after looking into the roborock database, i didn't find something called token.
Did i miss it ? or is only on the Xiaomi app ?

**Comment by mitchins** [comment]
Just making sure we are on the same page did you use a wifi network with absolutely no internet connectivity when issuing the configure network CLI command with miio to the device?

**Comment by mitchins** [comment]
App Versions after 5.1 or something don’t contain tokens any longer.. please check the exact version and take this into account. 

I’ve also found the last legacy version that does have tokens to fail to login these days.

**Comment by daweizhangau** [comment]
@poofyteddy 
Yes, I mean the Xiaomi's Mijia app. But for some devices/app versions, token might not be in that database. 
https://github.com/Maxmudjon/com.xiaomi-miio/blob/master/docs/obtain_token.md

**Comment by phrogg** [comment]
Any updates here?

**Comment by mitchins** [comment]
> Any updates here? 

I’m not sure what can be done, xiaomi have stopped embedding tokens.
My setup still works, I think people might be failing to segregate their VLANs properly to follow my approach.
